http-mclient/test/Test/Network/HTTP/Mclient.hs

```diff
@@ -0,0 +1,219 @@
+module Test.Network.HTTP.Mclient (testTree) where
+
+import Control.Monad (when)
+import Control.Monad.Reader (asks)
+import Data.Word (Word8)
+import GHC.Enum (succ)
+import GHC.Records (HasField(..))
+import MIO.Core
+import MPrelude
+import Test.Tasty
+import Test.Tasty.HUnit ((@?=))
+
+import qualified Data.Aeson               as JSON
+import qualified Data.List                as List
+import qualified MIO.Log                  as Log
+import qualified Network.HTTP.Client      as HTTP
+import qualified Network.HTTP.MClient     as HTTP
+import qualified Network.HTTP.Types       as HTTP
+import qualified Network.Socket           as Network
+import qualified Network.Wai              as WAI
+import qualified Network.Wai.Handler.Warp as Warp
+import qualified Test.Tasty.HUnit         as Tasty
+import qualified UnliftIO
+import qualified UnliftIO.Concurrent      as UnliftIO
+
+data ExpectedResponse
+  = ConnectionFailure
+  | ResponseTimeout
+  | Status500
+  | Success
+  deriving stock (Eq, Show)
+
+data Environment = Environment
+  { httpSendRequest :: HTTP.SendRequest
+  , logAction       :: Log.Action
+  , requestsCounter :: UnliftIO.MVar Word8
+  }
+
+testTree :: TestTree
+testTree = testGroup "Network.HTTP.MClient"
+  [ testGroup "with a connection failure"
+      [ Tasty.testCase "return an HTTP connection failure after a single retry" .
+          runWithEnvironment $ do
+            (port, socket)  <- liftIO Warp.openFreePort
+
+            -- assert zero requests
+            assertRequestsCount 0
+
+            UnliftIO.withAsync (runServer port socket) $ \server -> do
+              UnliftIO.withAsync (sendHttpRequest port) $ \client -> do
+                -- close server socket right before processing request
+                liftIO $ Network.close socket
+
+                assertResponse ConnectionFailure =<< UnliftIO.wait client
+
+                -- assert 2 requests, where the second is a retry
+                assertRequestsCount 2
+
+                UnliftIO.cancel server
+      ]
+  , testGroup "with response timeout"
+      [ Tasty.testCase "return an HTTP response timeout exception with no retries" .
+          withServer application $ \port -> do
+            assertRequestsCount 0
+
+            UnliftIO.withAsync (sendHttpRequest' port setRequestTimeout) $ \client -> do
+
```

Comment:
__suggestion: (nitpick)__ Is it typical to have a blank line after a `do` statement? I don't usually see this convention elsewhere.

Reply:
@dkubb not convention, I will remove it.